### PR TITLE
hda: chain transition states fix

### PR DESCRIPTION
**Pull request by makarukp**
Require maintenance in logic to handle multiple start/pause requests by one chain instance.

Signed-off-by: Piotr Makaruk <piotr.makaruk@intel.com>

### CODE REVIEW

**Comment by ujfalusi** [comment]
Created a draft PR with this and a revert to enable the ChainDMA on tgl: #6984

File: src/audio/chain_dma.c [hunk 1]
```diff
@@ -333,6 +240,114 @@ static enum task_state chain_task_run(void *data)
...
+	key = k_spin_lock(&drivers->lock);
+	switch (cd->chain_task.state) {
+	case SOF_TASK_STATE_QUEUED:
+		k_spin_unlock(&drivers->lock, key);
+		return 0;
+	case SOF_TASK_STATE_COMPLETED:
+		break;
+	case SOF_TASK_STATE_INIT:
+		break;
+	case SOF_TASK_STATE_FREE:
```
Comment (jsarha) [inline src/audio/chain_dma.c:261]: Is this addition really necessary, after the task initialization is properly done?

File: src/audio/chain_dma.c [hunk 2]
```diff
@@ -333,6 +240,114 @@ static enum task_state chain_task_run(void *data)
...
+		if (ret)
+			goto error;
+		ret = chain_host_start(dev);
+		if (ret) {
+			chain_link_stop(dev);
+			goto error;
+		}
+	}
+
+	pm_policy_state_lock_get(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
```
Comment (jsarha) [inline src/audio/chain_dma.c:289]: I can not find a symmetric pm_policy_state_lock_put() from the successful code path. Is it handled somewhere out side this function, or is it missing?

**Review by jsarha — COMMENTED** [review]
This is a definite improvement over the situation before this PR, since before this a resume after pause did not work at all. Now it works, at least most of the time. If I run "sof-test/test-case/multiple-pause-resume.sh -r 50" on my upx-i11 it fails more often than succeeds. However, hitting the failure usually takes quite a while. When looking at FW logs, I can see several:
[   16.733930] <err> copier: comp:0 0x40000 failed to find dai comp or sink pipeline not running.
and even more:
[   20.268613] <wrn> dai_comp: comp:1 0x40001 dai_copy(): nothing to copy

**Comment by jsarha** [comment]
I tried adding "pm_policy_state_lock_put(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);" after "cd->chain_task.state = SOF_TASK_STATE_INIT;" but it does not appear to help (nor make the behavior worse. One example of failure in the test logs: 

(3/50) pcm'Deepbuffer HDA Analog' cmd'aplay' id'5': Wait for 44 ms before resume
(3/50) pcm'HDA Analog' cmd'aplay' id'3': Wait for 49 ms before resume
(4/50) pcm'HDA Analog' cmd'aplay' id'3': Wait for 49 ms before pause        
(4/50) pcm'Deepbuffer HDA Analog' cmd'aplay' id'5': Wait for 48 ms before pause
(4/50) pcm'Deepbuffer HDA Analog' cmd'aplay' id'5': Wait for 49 ms before resume
(4/50) pcm'HDA Analog' cmd'aplay' id'3': Wait for 42 ms before resume
 aplay: pcm_write:2086: write error: Input/output error                     
aplay: pcm_write:2086: write error: Input/output error

2023-01-23 22:38:09 UTC [REMOTE_INFO] pipeline: HDA Analog with aplay
2023-01-23 22:38:09 UTC [REMOTE_INFO] pipeline: Deepbuffer HDA Analog with aplay
2023-01-23 22:38:09 UTC [REMOTE_INFO] Check expect exit status
declare -- cmd="journalctl_cmd --since=@1674513395"
2023-01-23 22:38:09 UTC [REMOTE_ERROR] pause resume PID 5752 had non-zero exit status

File: src/audio/chain_dma.c [hunk 3]
```diff
@@ -333,6 +240,114 @@ static enum task_state chain_task_run(void *data)
...
+
+	ret = schedule_task_init_ll(&cd->chain_task, SOF_UUID(chain_dma_uuid),
```
[the remark this answers is not in the document]
Comment (makarukp) [reply src/audio/chain_dma.c:291]: First of all complete initialization. `chedule_task_free` in `chain_task_pause` should matched with `schedule_task_init_ll` in `chain_task_start`. `pm_policy state_lock_get` is should enable prevent power gating and shall be triggered also in `chain_task_start` but could be also done after success in schedule_task for less complexity.

File: src/audio/chain_dma.c (hunk 1 above)
Comment (makarukp) [reply src/audio/chain_dma.c:261]: After stream resuming it has `SOF_TASK_STATE_FREE`, so we should tret such case the same as `SOF_TASK_STATE_INIT`.

File: src/audio/chain_dma.c (hunk 2 above)
Comment (makarukp) [reply src/audio/chain_dma.c:289]: `pm_policy_state_lock_put()` symmetric from successful code path can be found in at the end of `chain_task_pause` in under label called `error:`.

File: src/audio/chain_dma.c [hunk 4]
```diff
@@ -333,6 +240,114 @@ static enum task_state chain_task_run(void *data)
...
+	if (!ret)
+		ret = ret2;
+	if (ret < 0)
+		goto error;
+
+	cd->chain_task.state = SOF_TASK_STATE_COMPLETED;
+	schedule_task_free(&cd->chain_task);
+
+error:
+	pm_policy_state_lock_put(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
```
Comment (lyakh) [inline src/audio/chain_dma.c:346]: so in error case you don't free the task, but you do put the PM policy state lock, what's the logic behind this? BTW, reviewing this would've been easier is you kept functions in place, then differences could be seen directly

Comment (makarukp) [reply src/audio/chain_dma.c:341]: Moving `schedule_task_init` to `chain_task_start()` caused some undesired code rearrangement.

Comment (dabekjakub) [reply src/audio/chain_dma.c:340]: Agree with lyakh that if possible this should be done in a way that makes it more reviewable. Can't the commit be split into smaller pieces that would show differences better ?

Comment (makarukp) [reply src/audio/chain_dma.c:340]: I made additional preparation commit.

**Review by jsarha — APPROVED** [review]
Now after over 20 test runs of multiple-pause-resume I now saw an error also from the FW build where I had 6983 & 6656 reverted. So I guess its reasonable assume 6656 with 6983 is Ok. My review comments were also well explained. Approved.

File: src/audio/chain_dma.c [hunk 5]
```diff
@@ -284,17 +286,29 @@ static int chain_task_start(struct comp_dev *dev)
...
+	}
 
 	ret = schedule_task(&cd->chain_task, 0, 0);
-	if (ret) {
-		chain_host_stop(dev);
-		chain_link_stop(dev);
-		goto error;
+	if (ret < 0) {
+		comp_err(dev, "chain_task_start(), ll schedule task failed");
+		goto error_task;
```
Comment (lyakh) [inline src/audio/chain_dma.c:300]: how about this case - will the task be free on this path or not? Looks like it has to be freed here

**Review by lgirdwood — APPROVED** [review]
I will rerun CI after some regression fixes are merged today.

**Comment by marc-hb** [comment]
SOFCI TEST

**Comment by marc-hb** [comment]
`topology/topology2/avs-tplg/sof-nocodec-bt-lbm.tplg` build failure in https://sof-ci.01.org/sofpr/PR6983/build3565/build/tools.txt , https://sof-ci.01.org/sofpr/PR6983/build3565/build https://github.com/thesofproject/sof/actions/runs/4023985041/jobs/6915487545 and many others fixed by #6999

**Comment by marc-hb** [comment]
@makarukp to fix Github Actions and retest with #6999 you must perform an empty `git commit --amend -C HEAD` to change the SHA1 and then force-push.

Don't rebase as this makes it very difficult to see that nothing changed. Most CIs perform a merge before testing and our Github Actions do.

**Review by tmleman — APPROVED** [review]

**Comment by kv2019i** [comment]
@lyakh  good to go?
